File: dbTests/dbTests.cs
using System;
using System.IO;
using System.Collections.Generic;

using Xunit;

using Zumero.LSM;
using Zumero.LSM.fs;
using System.Threading;
using System.Threading.Tasks;
using lsm_tests;

namespace newTests
{
	public class MyClass
	{
		private static string tid()
		{
			return Guid.NewGuid ().ToString ().Replace ("{", "").Replace ("}", "").Replace ("-", "");
		}

		[Fact]
		public void empty_cursor()
		{
			var f = new dbf (tid());
			var db = new Zumero.LSM.fs.Database (f) as IDatabase;
			var csr = db.OpenCursor ();
			csr.First ();
			Assert.False (csr.IsValid ());
			csr.Last ();
			Assert.False (csr.IsValid ());
		}

		[Fact]
		public void several_threads()
		{
			var f = new dbf (tid());
			using (var db = new Zumero.LSM.fs.Database (f) as IDatabase) {
				var ta = new Thread[5];
				var ts = new Guid[ta.Length];

				ta[0] = new Thread(() => {
					var t1 = new Dictionary<byte[],Stream>();
					for (int i=0; i<5000; i++) {
						t1.Insert((i*2).ToString(), i.ToString());
					}
					ts[0] = db.WriteSegment (t1);
				});

				ta[1] = new Thread(() => {
					var t1 = new Dictionary<byte[],Stream>();
					for (int i=0; i<5000; i++) {
						t1.Insert((i*3).ToString(), i.ToString());
					}
					ts[1] = db.WriteSegment (t1);
				});

				ta[2] = new Thread(() => {
					var t1 = new Dictionary<byte[],Stream>();
					for (int i=0; i<5000; i++) {
						t1.Insert((i*5).ToString(), i.ToString());
					}
					ts[2] = db.WriteSegment (t1);
				});

				ta[3] = new Thread(() => {
					var t1 = new Dictionary<byte[],Stream>();
					for (int i=0; i<5000; i++) {
						t1.Insert((i*7).ToString(), i.ToString());
					}
					ts[3] = db.WriteSegment (t1);
				});

				ta[4] = new Thread(() => {
					var t1 = new Dictionary<byte[],Stream>();
					for (int i=0; i<5000; i++) {
						t1.Insert((i*11).ToString(), i.ToString());
					}
					ts[4] = db.WriteSegment (t1);
				});

				foreach (Thread t in ta) {
					t.Start();
				}

				foreach (Thread t in ta) {
					t.Join();
				}

				using (var tx = db.RequestWriteLock ()) {
					tx.PrependSegments (ts);
				}

				using (var csr = db.OpenCursor ()) {
					csr.First ();
					int count = 0;
					while (csr.IsValid ()) {
						count++;
						csr.Next ();
					}
					//Assert.Equal (20000, count);
				}
			}
		}

		[Fact]
		public void first_write()
		{
			var f = new dbf (tid());
			using (var db = new Zumero.LSM.fs.Database (f) as IDatabase) {
				// TODO consider whether we need IDatabase at all.  only
				// if we're going to do a C# version too, right?

				// build our data to be committed in memory in a
				// standard .NET dictionary.  note that are not
				// specifying IComparer here, so we're getting
				// reference compares, which is not safe unless you
				// know you're not inserting any duplicates.
				var mem = new Dictionary<byte[], Stream> ();
				for (int i = 0; i < 100; i++) {
					// extension method for .Insert(string,string)
					mem.Insert (i.ToString (), i.ToString ());
				}
				
				// write the segment to the file.  nobody knows about it
				// but us.  it will be written to the file, but the header
				// will not be touched yet, so its pages will be reclaimed
				// later unless this segment becomes official as a member
				// of the current state, listed in the header.
				var seg = db.WriteSegment (mem);

				// open a tx and add our segment to the current state.
				// after we do this, the segment is real.  "committed".
				using (var tx = db.RequestWriteLock ()) {
					var a = new List<Guid> { seg };
					tx.PrependSegments (a);
				}

				// open a cursor on the db and see if stuff looks okay
				using (var csr = db.OpenCursor ()) {
					csr.Seek (42.ToString (), SeekOp.SEEK_EQ);
					Assert.True (csr.IsValid ());
					csr.Next ();
					Assert.True (csr.IsValid ());
					string k = csr.Key ().UTF8ToString ();
					Assert.Equal ("43", k);
				}
			}
		}
	}
}

File: fsTests/dbTests.fs

module fsTests

open System
open System.Collections.Generic
open System.IO
open Xunit

open Zumero.LSM
open Zumero.LSM.fs

let tid() = 
    let g = Guid.NewGuid().ToString()
    let g2 = g.Replace ("{", "")
    let g3 = g2.Replace ("}", "")
    let g4 = g3.Replace ("-", "")
    g4

type dseg = Dictionary<byte[],Stream>

let to_utf8 (s:string) =
    System.Text.Encoding.UTF8.GetBytes (s)

let from_utf8 (ba:byte[]) =
    System.Text.Encoding.UTF8.GetString (ba, 0, ba.Length)

let insert (ds:dseg) (sk:string) (sv:string) =
    let k = to_utf8 sk
    let v = new MemoryStream(System.Text.Encoding.UTF8.GetBytes (sv))
    ds.[k] <- v

[<Fact>]
let empty_cursor() = 
    let f = dbf("empty_cursor" + tid())
    use db = new Database(f) :> IDatabase
    use csr = db.OpenCursor()
    csr.First ()
    Assert.False (csr.IsValid ())
    csr.Last ()
    Assert.False (csr.IsValid ())

[<Fact>]
let simple_write() = 
    let f = dbf("simple_write" + tid())
    use db = new Database(f) :> IDatabase
    let d = dseg()
    for i in 1 .. 100 do
        let s = i.ToString()
        insert d s s
    let seg = db.WriteSegment d
    async {
        use! tx = db.RequestWriteLock2()
        tx.CommitSegments [ seg ]
    } |> Async.RunSynchronously
    use csr = db.OpenCursor()
    csr.Seek ((42).ToString() |> to_utf8, SeekOp.SEEK_EQ)
    Assert.True (csr.IsValid())
    csr.Next()
    let k = csr.Key() |> from_utf8
    Assert.Equal<string> ("43", k)

[<Fact>]
let multiple() = 
    let f = dbf("multiple" + tid())
    use db = new Database(f) :> IDatabase
    let NUM = 10
    let rand = Random()

    let createMemorySegment (rand:Random) count =
        let d = Dictionary<byte[],Stream>()
        for q in 1 .. count do
            let sk = rand.Next().ToString()
            let sv = rand.Next().ToString()
            insert d sk sv
        d

    let start i = async {
        let commit g = async {
            use! tx = db.RequestWriteLock2()
            tx.CommitSegments (g :: List.empty)
            if i%4=0 then
                match db.MergeAll2() with
                | Some f ->
                    let blk = async {
                        let! g = f
                        tx.CommitMerge g
                    }
                    do! blk
                | None -> ()
            }

        let count = rand.Next(10000)
        let d = createMemorySegment rand count
        let g = db.WriteSegment(d)
        do! commit g
    }

    let c = seq { for i in 0 .. NUM-1 do yield i; done }
    let workers = Seq.fold (fun acc i -> (start i) :: acc) List.empty c
    let go = Async.Parallel workers
    Async.RunSynchronously go |> ignore

    match db.MergeAll2() with
    | Some f ->
        async {
            let! g = f
            use! tx = db.RequestWriteLock2()
            tx.CommitMerge g
        } |> Async.RunSynchronously
    | None -> ()

    let loop() = 
        use csr = db.OpenCursor()
        csr.First()
        let mutable count = 0
        while csr.IsValid() do
            count <- count + 1
            csr.Next()

    loop()

[<Fact>]
let lexographic() = 
    let f = dbf("lexographic" + tid())
    use db = new Database(f) :> IDatabase
    let d = dseg()
    insert d "8" ""
    insert d "10" ""
    insert d "20" ""
    let g = db.WriteSegment(d)
    async {
        use! tx = db.RequestWriteLock2()
        tx.CommitSegments [ g ]
    } |> Async.RunSynchronously

    use csr = db.OpenCursor()
    csr.First()
    Assert.True(csr.IsValid())
    Assert.Equal<string> ("10", csr.Key () |> from_utf8)

    csr.Next()
    Assert.True(csr.IsValid())
    Assert.Equal<string> ("20", csr.Key () |> from_utf8)

    csr.Next()
    Assert.True(csr.IsValid())
    Assert.Equal<string> ("8", csr.Key () |> from_utf8)

    csr.Next()
    Assert.False(csr.IsValid())

    // --------
    csr.Last()
    Assert.True(csr.IsValid())
    Assert.Equal<string> ("8", csr.Key () |> from_utf8)

    csr.Prev()
    Assert.True(csr.IsValid())
    Assert.Equal<string> ("20", csr.Key () |> from_utf8)

    csr.Prev()
    Assert.True(csr.IsValid())
    Assert.Equal<string> ("10", csr.Key () |> from_utf8)

    csr.Prev()
    Assert.False(csr.IsValid())
